src-electron/generator/helper-zcl.js

```diff
@@ -2693,6 +2698,293 @@ function if_compare(leftValue, rightValue, options) {
   }
 }
 
+/**
+ *
+ * @param {*} type
+ * @param {*} dataType
+ * @param {*} clusterId
+ * @param {*} packageIds
+ * @param {*} context
+ * @returns The size and sign of a zcl data type
+ */
+async function zcl_data_type_size_and_sign(
+  type,
+  dataType,
+  clusterId,
+  packageIds,
+  context
+) {
+  let result = 0
+  let isSigned = false
+  if (dataType.discriminatorName.toLowerCase() == dbEnum.zclType.bitmap) {
+    let bitmap = await queryZcl.selectBitmapByNameAndClusterId(
+      context.global.db,
+      dataType.name,
+      clusterId,
+      packageIds
+    )
+    result = bitmap.size
+  } else if (dataType.discriminatorName.toLowerCase() == dbEnum.zclType.enum) {
+    let en = await queryZcl.selectEnumByNameAndClusterId(
+      context.global.db,
+      dataType.name,
+      clusterId,
+      packageIds
+    )
+    result = en.size
+  } else if (
+    dataType.discriminatorName.toLowerCase() == dbEnum.zclType.number
+  ) {
+    let number = await queryZcl.selectNumberByNameAndClusterId(
+      context.global.db,
+      dataType.name,
+      clusterId,
+      packageIds
+    )
+    isSigned = number.isSigned
+    result = number.size
+  } else {
+    env.logWarning(type + ' is a complex type and size could not be determined')
+  }
+  return { size: result, isSigned: isSigned }
+}
+
+/**
+ * Available Options:
+ * - asUpperCamelCase: 0/1 to return cpp data type starting with lower/upper
+ * camel case
+ * @param {*} type
+ * @param {*} dataType
+ * @param {*} hash
+ * @param {*} context
+ * @returns The corresponding cpp data type for a zcl data type.
+ */
+async function as_underlying_objective_c_zcl_type(
+  type,
+  dataType,
+  hash,
+  context
+) {
+  let res = await new Promise((resolve, reject) => {
+    let asUpperCamelCase =
+      hash && hash.asUpperCamelCase ? hash.asUpperCamelCase : false
+    if (dataType.discriminatorName.toLowerCase() == dbEnum.zclType.string) {
+      if (octetStringTypes.includes(type.toUpperCase())) {
+        resolve('OctetString')
```

Comment:
There is _definitely_ no objective C OctetString/CharString thing.

Reply:
This comes from attribute_data_callback_name which is referenced by MTRBaseClusters-src.zapt. I named it objectiveC because it was used to generate a .mm file. Do you want me to name this something else?

Reply:
I don't see any OctetString bits in https://github.com/project-chip/connectedhomeip/blob/3c6bed3bb28f66cc2e1ebbe3561fa0838f6b7c51/src/darwin/Framework/CHIP/templates/partials/attribute_data_callback_name.zapt (as of today).  What am I missing?

Reply:
https://github.com/project-chip/connectedhomeip/blob/3c6bed3bb28f66cc2e1ebbe3561fa0838f6b7c51/src/darwin/Framework/CHIP/templates/partials/attribute_data_callback_name.zapt#L18 that part?

Reply:
I guess it's doing `chipCallback.name`?

If that's the thing we are trying to deal with this has nothing to do with "as_underlying_objective_c_zcl_type".  What's really going on there is how the type name is being mapped into function type names (for the callback function types).  The actual setup there is somewhat complicated, since it needs to encode nullability, lists of things (just bare "List" is definitely not ok there, for example).

Reply:
I see that there are more references to `chipCallback.name` than I had pulled in locally but I was able to handle that using the following in attribute_data_callback_name.zapt:
`{{as_underlying_language_specific_zcl_type type ../id signedPrefix='Int' unSignedPrefix='Int' unSignedPostfix='u' signedPostfix='s' sizeIn='bits' roundUpToPowerOfTwo=1 language='objectiveC' boolean='Boolean' asUpperCamelCase=1}}`

I now understand that I should not be using `language='objectiveC'` here. I can move all of that into a helper with less options and call it `{{as_underlying_callback_for_zcl_type type ../id}}` . I could also continue to have those options if that needs to be different at times based on language being generated for(Make it more scalable).

Reply:
The thing is, as_underlying_callback_for_zcl_type doesn't really make sense either, since it's only (1) part of the callback name (2) only for some types.

What does that function end up returning in the end?  Maybe we can just replace it with something simpler....

Reply:
The function as_underlying_language_specific_zcl_type/as_underlying_callback_for_zcl_type just returns a string based on the zcl type given

Reply:
Right now it is doing ```async function as_underlying_objective_c_zcl_type(
  type,
  dataType,
  hash,
  context
) {
  let res = await new Promise((resolve, reject) => {
    let asUpperCamelCase =
      hash && hash.asUpperCamelCase ? hash.asUpperCamelCase : false
    if (dataType.discriminatorName.toLowerCase() == dbEnum.zclType.string) {
      if (octetStringTypes.includes(type.toUpperCase())) {
        resolve('OctetString')
      } else if (characterStringTypes.includes(type.toUpperCase())) {
        resolve('CharString')
      } else if (context.isArray) {
        resolve('List')
      } else {
        resolve(null)
      }
    } else if (dataType.name.includes('float')) {
      return asUpperCamelCase ? resolve('Float') : resolve('float')
    } else if (dataType.name.includes('double')) {
      return asUpperCamelCase ? resolve('Double') : resolve('double')
    } else if (dataType.name.includes('single')) {
      return asUpperCamelCase ? resolve('Float') : resolve('float')
    }
    resolve(null)
  })
  return res
}```

Reply:
For now I renamed the helper to avoid any language specific activity. Instead it looks for a modifier such as `chipCallbackName` to determine the right output for exceptions.

Reply:
`as_underlying_zcl_type_modifier` helper handles `chipCallbackName`. The above image was not the complete implementation. `as_underlying_zcl_type_modifier` calls the above image along the way to handle certain exceptions. If the above image returns null then it handles integers like you asked.

Reply:
> Regarding the context.isArray. I believe I got that fromClusterHelper#asChipCallback

Yes, but we never use `chipCallback.name` if `isArray` is true, at least in the Darwin code.

So just so I can understand: all this machinery is to replace the `name` computation we have right now in `asChipCallback`?  It looks like that's used in one place in the Darwin code and in a bunch of places in Java code.

Speaking about the Darwin side for now, and looking at asChipCallback:

1. We can detect octet and char strings directly in the template and do the right thing.
2. We already detect arrays in the template.
3. For the numeric types, what we really need are:
    a. Ways to detect/canonicalize the type for bool/float/double.  Do those exist already?
    b. A way to get the integer width, in bytes, for an integer type.
    c. A way to check whether an integer type is signed, for an integer type.

If we had those pieces for numeric types, this is all doable in the template without building all this machinery.

I can't speak to the java pieces, though even there I expect a lot of it can be done similarly.

Basically, I don't think we should be building a 1-1 replacement for `chipCallback.name`; that's a very leaky abstraction that sort of made sense when the helpers were in-tree but does not make sense as an actual ZAP API.  Instead we should be exposing information about the types and letting people write templates that do what they want.

Reply:
I am a proponent of doing things in the template but have had resistance with adding too much to the template in the past. There will be a lot of if else logic in the template. Are we ok with that? If so, then I can remove the logic surrounding `chipCallback.name`
The java code has a lot more references to `chipCallback.name`

Reply:
Are you ok with doing the following in the template instead
`{{#if_is_octet_string type}}
      OctetString
      {{else}}
        {{#if_is_char_string type}}
        CharString
        {{else}}
          {{#if isArray}}
          List
          {{else if (is_str_equal type "float")}}
          Float
          {{else if (is_str_equal type "double")}}
          Double
          {{else if (is_str_equal type "single")}}
          Float
          {{else if (is_str_equal type "boolean")}}
          Boolean
          {{else}}
            {{#if_is_data_type_signed type ../id}}
            Int{{as_zcl_data_type_size type ../id sizeIn='bits' roundUpToPowerOfTwo=1}}s
            {{else}}
            Int{{as_zcl_data_type_size type ../id sizeIn='bits' roundUpToPowerOfTwo=1}}u
            {{/if_is_data_type_signed}}
          {{/if}}
        {{/if_is_char_string}}
      {{/if_is_octet_string}}`

Reply:
> if_is_octet_string type

Note that we have `isOctetString` already.

> if_is_char_string

Similarly, we have `isCharString`

> `{{#if isArray}} List`

That's not what the template would have, but sure.

> `is_str_equal type "float"`

An `isFloat` that works on types and does the float/single thing would be useful, perhaps, but in practice do we actually have both "float" and "single" in the XML?   Alternately, the template could just do `if (or ....)` as needed.

> if_is_data_type_signed

That looks great, but why does it need `../id`?

> as_zcl_data_type_size

That also looks great.  We could simplify things a bit by doing the "Int{size}" bits once and just have the signed conditional around the "s/u" decision.

But yes, in general that sort of if/else cascade is just fine and we do that in other places (e.g. `src/darwin/Framework/CHIP/templates/partials/init_struct_member.zapt`, `src/darwin/Framework/CHIP/templates/partials/decode_value.zapt`, etc.

Reply:
ok I will start adding the requirements for the above.
`../id` is the clusterId in this case.

Reply:
Fixed this. Got rid of it.